cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ApplicationLifecycleService.java

```diff
@@ -896,7 +907,12 @@ private ApplicationWithPrograms deployApp(NamespaceId namespaceId, @Nullable Str
       throw new InvalidArtifactException(String.format("No application class found in artifact '%s' in namespace '%s'.",
                                                        artifactDetail.getDescriptor().getArtifactId(), namespaceId));
     }
-
+    Set<String> capabilities = appClass.getRequirements().getCapabilities();
+    for (String capability : capabilities) {
+      if (!capabilityReader.isEnabled(capability)) {
```

Comment:
You can simplify it to

```
boolean allEnabled = appClass.getRequirements().getCapabilities().stream().allMatch(capabilityReader::isEnabled);
if (!allEnabled) {
  throw ...
}
```

Reply:
didnt use stream API because of needing to wrap IOException as a RuntimeException